### PR TITLE
Added support for M3 filled and filled tonal buttons.

### LINKED ISSUE #104357: [Proposal] Add filled buttons as per material 3 guidelines
# Proposal

Add factories on ElevatedButton for FilledButton and FilledTonalButton or add factories to an existing button to respect m3 guidelines. I understand those were not added because then what about when the background needs to be an error background ? tertiary ? etc. I believe however that a few factories could help here.

relevant:   https://github.com/flutter/flutter/issues/99022  

# use case

Going from a design using the material 3 kit on Figma to flutter should be seamless.

### PR DESCRIPTION
**Pull request by darrenaustin**
Followup to #100794.

This adds support for the filled and filled tonal [Material Design 3 buttons](https://m3.material.io/components/buttons/overview):

Screen Shot 2022-07-10 at 5 33 28 PM

To create a filled button use the new `FilledButton` and to create a tonal variant, use `FilledButton.tonal`:

```dart
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(colorSchemeSeed: const Color(0xff6750a4), useMaterial3: true),
      home: Scaffold(
        appBar: AppBar(title: const Text('FilledButton Sample')),
        body: Center(
          child:  Column(
            children: <Widget>[
              FilledButton(
                onPressed: () {},
                child: const Text('Filled'),
              ),
              FilledButton.tonal(
                onPressed: () {},
                child: Text('Filled tonal'),
              ),
            ]
          ),
          // ...
```

Fixes: #104357

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/button_style_button.dart:25]: Consider adding [FilledButton.tonal]

File: packages/flutter/lib/src/material/elevated_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/elevated_button.dart:59]:
```suggestion
///  * [TextButton], a simple flat button without a fill or outline.
```

File: packages/flutter/lib/src/material/elevated_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/elevated_button.dart:57]:
```suggestion
///  * [OutlinedButton], a filled button whose material elevates when pressed.
```

For consistency

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:23]:
```suggestion
/// A Material Design filled button.
```

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:27]:
```suggestion
/// like **Save**, **Join now**, or **Confirm**.
```

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:60]: Add filledbutton.tonal

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:49]:
```suggestion
/// To create a 'filled tonal' buton, use [FilledButton.tonal].
```

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:84]:
```suggestion
  /// Create a filled button from [icon] and [label].
```

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:87]: I think we should move away from encode spec info in docs

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:32]: I'd remove this and point to [styleFrom] instead

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:130]: Consider removing this

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:418]: I think this is available in tokens, but not sure

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:486]: where do these values come from?

File: packages/flutter/lib/src/material/button_style.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/button_style.dart:68]: Is there somewhere in docs that explains how to achieve the tonal look with `FilledButton.styleFrom`?

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

Looks great! Minor comments about documentation and token usage

File: packages/flutter/lib/src/material/button_style.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/button_style.dart:68]: No. We used to have an example that showed how to do it with and ElevatedButton, but I updated it to just use FilledButton.tonal.

File: packages/flutter/lib/src/material/filled_button.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/filled_button.dart:192]: Please add example for FilledButton.tonal

File: packages/flutter/lib/src/material/button_style.dart
Comment (guidezpl) [reply packages/flutter/lib/src/material/button_style.dart:68]: Gotcha. I commented where I think docs would benefit from a small example.

File: examples/api/lib/material/filled_button/filled_button.0.dart
Comment (Reprevise) [inline examples/api/lib/material/filled_button/filled_button.0.dart:7]: Shouldn't this be FilledButton instead of ElevatedButton?

File: packages/flutter/lib/src/material/filled_button.dart
Comment (thipokch) [inline packages/flutter/lib/src/material/filled_button.dart:393]: Fyi, as per material 3 spec, opacity for this state is at 12%. This applies to other buttons too!
Reference
```suggestion
      return overlay.withOpacity(0.12);
```

File: packages/flutter/lib/src/material/filled_button.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/filled_button.dart]: Yup this was a left over from some backwards compatibility stuff from another button code that was copied. Fixed.

File: examples/api/lib/material/filled_button/filled_button.0.dart
Comment (darrenaustin) [reply examples/api/lib/material/filled_button/filled_button.0.dart:7]: D'oh! Missed in a copy/paste sweep. Thx!

File: packages/flutter/lib/src/material/filled_button.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/filled_button.dart:174]: [ButtonStyle].mouseCursor => [ButtonStyle.mouseCursor]

File: packages/flutter/lib/src/material/filled_button.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/filled_button.dart:190]: NICE

File: packages/flutter/lib/src/material/filled_button.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/filled_button.dart:509]: In methods like this you could avoid creating closures:
```dart
Color? getBackgroundColor(Set<MaterialState> states) {
  if (states.contains(MaterialState.disabled)) {
    return _colors.onSurface.withOpacity(0.12);
  }
  return _colors.primary;
}
MaterialStateProperty<Color?>? get backgroundColor => MaterialStateProperty.resolveWith(getBackgroundColor);
```

Also: shouldn't overrides like this return `MaterialStateProperty<Color?>`, not a nullable MaterialStateProperty?

File: packages/flutter/lib/src/material/filled_button.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/filled_button.dart:571]: We need a way to make values like this const. Could be a new issue; separate PR.

**Review by HansMuller — APPROVED** [review]
LGTM however there are many comments from @guidezpl that merit response/changes.

File: packages/flutter/lib/src/material/filled_button.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/filled_button.dart:509]: Is the concern here that we are generating short lived objects with the closure (and thus a potential performance issue), or is this just a style preference? I'll change it, but I am just curious.

And yes the override should be `MaterialStateProperty<Color?>?` to be consistent. I think this wasn't flagged as overrides are allowed to constrain the return value like this?

Comment (darrenaustin) [reply packages/flutter/lib/src/material/filled_button.dart:509]: Actually, this is turning into a large change, so I will do this closure removal for all the templates in another PR.

File: packages/flutter/lib/src/material/filled_button.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/filled_button.dart]: Unfortunately these are a little harder to pull out of the database in the current button template (which is used for all buttons, some of which may not have an opacity for each of these). I will follow up on this in a future PR.

File: packages/flutter/lib/src/material/filled_button.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/filled_button.dart:458]: Copied from the other ButtonStyle classes 😄? That said, I don't think these padding values are in the tokens yet. When they are we will have to do a sweep across the widgets to include the layout data. Currently these are hard coded.

File: packages/flutter/lib/src/material/filled_button.dart
Comment (HansMuller) [reply packages/flutter/lib/src/material/filled_button.dart:516]: That sounds good. I suggested it because there should be a small performance benefit.

**Comment by rydmike** [comment]
This will also close https://github.com/flutter/flutter/issues/104357, hope it lands in master soon so it can make it to next stable release.

File: packages/flutter/lib/src/material/theme_data.dart
Comment (albertms10) [inline packages/flutter/lib/src/material/theme_data.dart:1188]: Whitespace typo (?) 👀

```suggestion
  ///   * Common buttons: [ElevatedButton], [FilledButton], [OutlinedButton], [TextButton]
```

Comment (darrenaustin) [reply packages/flutter/lib/src/material/theme_data.dart]: Ooops. Fixed. Thx.

**Comment by darrenaustin** [comment]
> This will also close #104357, hope it lands in master soon so it can make it to next stable release.

Sorry this has taken a while. There is an internal Google naming conflict that we need to resolve before we can land this (as well as I was out for a couple weeks vacation). Hopefully this should land in the next couple of days.

**Comment by rydmike** [comment]
No worries @darrenaustin, I suspect this one won't make it to next stable. It looked so ready that I just thought it would be a pity to wait for it another 3-4 months for the next release after the assumed imminent one. However, there are so many other feature gaps and issues with M3 opt-in in Beta 3.3.0-x and even master, that I can't recommend opting in on M3 for Flutter projects yet anyway. This one was rather small from that point of view, since you can compose/configure this style already when needed.

**Comment by hoanghuynh296** [comment]
Great work! Many thanks!

**Pull request merged**
